### ISSUE TITLE
Market makers use a p2sh address with some probability

### ISSUE THREAD

**Issue by chris-belcher** [chris-belcher/joinmarket#84]
It would provide some privacy for senders to p2sh. The actual form of p2sh could be a mixture of 2-of-2, 2-of-3 and perhaps other types. Passive observers of the blockchain don't know the private keys are all owned by the same entity.

I think 2-of-3 is the most common, although sites like http://p2sh.info/ don't show that specific statistic.

**Comment by adlai** [comment]
Two thoughts:
1. This adds significant complexity to the wallet. It's no longer sufficient to store a single index, or a single gap limit; there needs to be some way of knowing which address combinations to check for p2sh. One possibility is to use a separate BIP32 branch for each address type, for example (M representing an arbitrary mixdepth):
  * m/0/M/0/### - P2PKH, incoming
  * m/0/M/1/### - P2PKH, change
  * m/0/M/2/### - 2of2 P2SH, incoming
    * m/0/M/2/even - 1st pubkey
    * m/0/M/2/odd - 2nd pubkey
  * m/0/M/3/### - 2of2 P2SH, change
    * m/0/M/3/even - 1st pubkey
    * m/0/M/3/odd - 2nd pubkey
  * m/0/M/4/### - 2of3 P2SH, incoming
    * m/0/M/4/0mod3 - 1st pubkey
    * m/0/M/4/1mod3 - 2nd pubkey
    * m/0/M/4/2mod3 - 3rd pubkey
  * m/0/M/5/### - 2of3 P2SH, change
    * m/0/M/5/0mod3 - 1st pubkey
    * m/0/M/5/1mod3 - 2nd pubkey
    * m/0/M/5/2mod3 - 3rd pubkey
2. A market maker running off RPC should pick address type according to popularity over the last N blocks.

**Comment by the9ull** [comment]
This would increase the size of the tx.
What about use p2sh only when p2sh requested by the taker? S/he could request (N-K) P2SH outputs and P2PKH outputs

**Comment by chris-belcher** [comment]
CoinJoins in general make the transaction bigger, it might end up being something we simply accept.
But your idea could also be a viable solution too, yes.

What happens when each of those makers wants to spend their p2sh coins under the taker-requests idea? Does it draw any more attention to the transaction.

**Comment by the9ull** [comment]
> What happens when each of those makers wants to spend their p2sh coins under the taker-requests idea? Does it draw any more attention to the transaction.

I don't think that it is an issue when p2sh are used as inputs, this does not reveal anything of unknown.
